crates/astria-sequencer/src/service/mempool.rs

```diff
@@ -280,21 +208,24 @@ async fn handle_check_tx<S: accounts::StateReadExt + address::StateReadExt + 'st
         Ok(nonce) => nonce,
     };
 
-    let actions_count = signed_tx.actions().len();
-
-    mempool
-        .insert(signed_tx, current_account_nonce)
+    if let Err(err) = mempool
+        .insert(the_tx.clone(), current_account_nonce)
         .await
-        .expect(
-            "tx nonce is greater than or equal to current account nonce; this was checked in \
-             check_nonce_mempool",
-        );
+        .context("mempool rejected validated transaction")
+    {
+        return response::CheckTx {
+            code: AbciErrorCode::INTERNAL_ERROR.into(),
+            info: AbciErrorCode::INTERNAL_ERROR.to_string(),
+            log: format!("transaction failed execution because: {err:#?}"),
+            ..response::CheckTx::default()
+        };
+    }
     let mempool_len = mempool.len().await;
 
     metrics
         .record_check_tx_duration_seconds_insert_to_app_mempool(finished_check_removed.elapsed());
-    metrics.record_actions_per_transaction_in_mempool(actions_count);
-    metrics.record_transaction_in_mempool_size_bytes(tx_len);
+    metrics.record_actions_per_transaction_in_mempool(the_tx.actions().len());
+    metrics.record_transaction_in_mempool_size_bytes(bytes.len());
     metrics.set_transactions_in_mempool_total(mempool_len);
 
     response::CheckTx::default()
```

Comment:
This file is really undertested in general. I think that at least transactions that fail execution due to too high of a nonce should be let into the mempool. There could be an argument that only stateful/stateless checks should be checked here and not execution, most mempools only filter against nonce and balance. We're different because we don't let failing transactions into blocks, but I'm a little concerned that requiring execution success is too strict.

Here's a test that should work with this PR showing the too high nonce failure: 

```suggestion
    response::CheckTx::default()
}
    
#[cfg(test)]
mod test {
    use std::sync::Arc;

    use astria_core::{
        crypto::{
            SigningKey,
            VerificationKey,
        },
        protocol::transaction::v1alpha1::{
            action::TransferAction,
            TransactionParams,
            UnsignedTransaction,
        },
        sequencer::{
            Account,
            AddressPrefixes,
            UncheckedGenesisState,
        },
    };
    use prost::Message as _;
    use rand::rngs::OsRng;
    use tendermint::abci::Code;

    use super::*;
    use crate::{
        app::test_utils::default_fees,
        mempool::Mempool,
        metrics::Metrics,
        test_utils::astria_address,
    };

    async fn new_setup(funded_key: Option<VerificationKey>) -> (Storage, Mempool) {
        let accounts = if funded_key.is_some() {
            vec![Account {
                address: crate::test_utils::astria_address(&funded_key.unwrap().address_bytes()),
                balance: 10u128.pow(19),
            }]
        } else {
            vec![]
        };
        let genesis_state = UncheckedGenesisState {
            accounts,
            address_prefixes: AddressPrefixes {
                base: crate::test_utils::ASTRIA_PREFIX.into(),
            },
            authority_sudo_address: crate::test_utils::astria_address(&[0; 20]),
            ibc_sudo_address: crate::test_utils::astria_address(&[0; 20]),
            ibc_relayer_addresses: vec![],
            native_asset_base_denomination: crate::test_utils::nria(),
            ibc_params: penumbra_ibc::params::IBCParameters::default(),
            allowed_fee_assets: vec!["nria".parse().unwrap()],
            fees: default_fees(),
        }
        .try_into()
        .unwrap();

        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mempool = Mempool::new();
        let metrics = Box::leak(Box::new(Metrics::new()));
        let mut app = App::new(snapshot, mempool.clone(), metrics).await.unwrap();
        app.init_chain(storage.clone(), genesis_state, vec![], "test".to_string())
            .await
            .unwrap();
        app.commit(storage.clone()).await;

        (storage.clone(), mempool)
    }

    #[tokio::test]
    async fn handle_check_tx_future_nonce_ok() {
        let signing_key_0 = SigningKey::new(OsRng);
        let signing_key_1 = SigningKey::new(OsRng);
        let (storage, mempool) = new_setup(Some(signing_key_0.verification_key())).await;

        let metrics = Box::leak(Box::new(Metrics::new()));

        // make valid transfer with higher nonce
        let value = 333_333;
        let tx = UnsignedTransaction {
            params: TransactionParams::builder()
                .nonce(1)
                .chain_id("test")
                .build(),
            actions: vec![
                TransferAction {
                    to: astria_address(&signing_key_1.address_bytes()),
                    amount: value,
                    asset: crate::test_utils::nria().into(),
                    fee_asset: crate::test_utils::nria().into(),
                }
                .into(),
            ],
        };

        let signed_tx = Arc::new(tx.into_signed(&signing_key_0));

        let request = request::CheckTx {
            tx: <astria_core::protocol::transaction::v1alpha1::SignedTransaction as Clone>::clone(
                &signed_tx,
            )
            .into_raw()
            .encode_to_vec()
            .into(),
            kind: request::CheckTxKind::New,
        };

        let res = handle_check_tx(request, storage, mempool, metrics).await;

        assert!(matches!(res.code, Code::Ok), "higher nonce should be ok");
    }
```

Reply:
Check-tx is now trying to downcast the errors to `InvalidNonce`: if `InvalidNonce` is found, and if it's due to the nonce being too far ahead, the transaction is still put into the mempool.